add docs for date module

dkirchhof opened the pull request.

zth commented:
@dkirchhof fantastic work! 👏 

Let me know when you want me to take a look at it.

dkirchhof commented:
@zth  Ready to go!

JS dates are a pain. I'm curious about the temporal api.

zth commented on src/Core__Date.resi:
```diff
@@ -35,16 +39,132 @@ type localeOptions = {
-@new external makeWithYMD: (~year: int, ~month: int, ~date: int) => t = "Date"
-@new external makeWithYMDH: (~year: int, ~month: int, ~date: int, ~hours: int) => t = "Date"
+/** 
+`valueOf(date)`
```
Once this is in main I'll actually remove this function. Not needed as we have `getTime` already and they're the same apparently.

zth commented on src/Core__Date.resi:
```diff
@@ -65,10 +191,27 @@ external makeWithYMDHMSM: (
-  @val external makeWithYM: (~year: int, ~month: int) => time = "Date.UTC"
-  @val external makeWithYMD: (~year: int, ~month: int, ~date: int) => time = "Date.UTC"
+  /** 
+  Same as Date.makeWithYM, but returns the time since 1 January 1970 UTC in milliseconds.
```
Could we add examples here too, preferably also comparing this with it's "regular" equivalent function?

And maybe a link to somewhere you can read about UTC vs regular dates? If such a thing exists on MDN.

zth reviewed (commented):
Great work, really solid! I have a few comments, but they would be mostly icing on the cake, this is already really great.

dkirchhof commented on src/Core__Date.resi:
````diff
@@ -35,16 +39,125 @@ type localeOptions = {
+// 1676851200000
+```
+*/
 @send external valueOf: t => time = "valueOf"
````
Sure :)

dkirchhof commented on src/Core__Date.resi:
```diff
@@ -35,16 +39,125 @@ type localeOptions = {
+Date.makeWithYM(~year=2023, ~month=12)
+// 2024-01-01T00:00:00.000Z
+
+Date.makeWithYM(~year=2023, ~month=-1)
```
I didn't know as well, but I had troubles to explain it. So I added the example :smile:

dkirchhof commented:
Thanks for the quick feedback. I will continue tomorrow ;)

zth commented:
> Thanks for the quick feedback. I will continue tomorrow ;)

Thank you for your contributions, it's great!

glennsl commented on src/Core__Date.resi:
```diff
@@ -2,7 +2,11 @@
+*/
 type t = Js.Date.t
+
 type time = float
```
Yes, `millisSinceEpoch` was the suggestion from @cknitt.

glennsl replied:
And to add an explanation in an associated docstring.

dkirchhof replied:
Would `msSinceEpoch` be okay?

glennsl replied:
I think so, at least.

zth commented on src/Core__Date.res:
```diff
@@ -1,6 +1,6 @@
 type t = Js.Date.t
 
-type time = float
+type msSinceEpoch = float
```
```suggestion
/**
Time, in milliseconds, since the UNIX epoch (January 1, 1970 00:00:00 UTC).
*/
type msSinceEpoch = float
```
cc @glennsl what do you think of this docstring?

glennsl replied:
Perfect! :smiley:

dkirchhof commented:
Should we mention that all the examples depend on the local time zone?
I was using `export TZ='Europe/London' && node test/TempTests.mjs` so the results of the examples wouldn't confuse anyone (my local time zone is MEZ).

zth commented:
> Should we mention that all the examples depend on the local time zone? I was using `export TZ='Europe/London' && node test/TempTests.mjs` so the results of the examples wouldn't confuse anyone (my local time zone is MEZ).

That's a good idea for sure, that can be slightly confusing with dates. I guess we'll need to add a "disclaimer" to all examples?

dkirchhof commented:
Changed the docs about epoch, added some more examples and added the disclaimer about results according to the local time zone.

zth commented:
> Changed the docs about epoch, added some more examples and added the disclaimer about results according to the local time zone.

This is looking great to me. You feel done with it?

dkirchhof commented:
> > Changed the docs about epoch, added some more examples and added the disclaimer about results according to the local time zone.
> 
> This is looking great to me. You feel done with it?

I would say so. Otherwise someone else could open a new PR for improvements?!

zth commented:
Looking great to me, thank you!
